include/range/v3/view/cycle.hpp

```diff
@@ -0,0 +1,259 @@
+                    : rng_(&rng), it_(ranges::begin(rng.rng_)), pos_(0)
+                {
+                }
+
+                reference current() const
+                {
+                    return *it_;
+                }
+
+                void next()
+                {
+                    ++pos_;
+                    const auto end = ranges::end(rng_->rng_);
+                    if (it_ != end)
+                    {
+                        ++it_;
+                    }
+                    if (it_ == end)
+                    {
+                        it_ = begin(rng_->rng_);
+                    }
+                }
+
+                bool equal(cursor const &pos) const
+                {
+                    return rng_ == pos.rng_ && it_ == pos.it_ && pos_ == pos.pos_;
+                }
+
+                CONCEPT_REQUIRES(BidirectionalRange<Rng>{})
+                void prev()
+                {
+                    --pos_;
+                    if (it_ == begin(rng_->rng_))
+                    {
+                        it_ = end(rng_->rng_);
+                    }
+                    --it_;
+                }
+
+                CONCEPT_REQUIRES(RandomAccessRange<Rng>())
+                void advance(difference_type n)
+                {
+                    if (n == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    range_difference_t<Rng> distance = ranges::distance(rng_->rng_);
+                    if (distance == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    pos_ += n;
+                    const auto begin = ranges::begin(rng_->rng_);
+                    const auto end = ranges::end(rng_->rng_);
+
+                    const difference_type pos = it_ - begin;
+                    const auto rel_distance = n >= 0 ? distance - pos : pos;
+
+                    if (n >= 0)
+                    {
+                        if (n < rel_distance)
+                        {
+                            it_ = begin + pos + n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + n % distance;
+                        }
+                    }
+                    else
+                    {
+                        n *= -1;
+                        if (n <= rel_distance)
+                        {
+                            it_ = begin + pos - n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + distance - n % distance;
+                            if (it_ == end)
+                            {
+                                it_ = begin;
+                            }
+                        }
+                    }
+                }
```

Comment:
@CaseyCarter Thanks! I have not removed `pos_` yet since I haven't made the tests to work with `take_exactly` yet but I'll keep trying. I've just committed your implementation of advance, is the copyright correct?

> Removing pos_ also allows you use the underlying range's difference_type directly.

I wasn't sure about this. The underlying range's `difference_type` might be very short, but through `view::cycle` which is infinite one might want to pass a much larger `n` to `advance`, so I thought the `difference_type` of the cursor should be `max<range_difference_t<Rng>, std::ptrdiff_t>` at least. 

> views have reference semantics - there should not be a non-const overload of `begin_cursor`/`end_cursor`.

Maybe I misunderstood views completely, but how do you handle mutation/non-mutation then? IIUC a const view provides `const_iterator`s, while a non-const view provides `iterator`s.